### PR TITLE
Fix some issues in the OptimizeTaskGroups optimizer

### PR DESCRIPTION
**Pull request by thinkharderdev**
Try and fix and cleanup the optimizer. I think it was not handling reduce tasks correctly and we mostly avoided nastiness by dumb luck....

### CODE REVIEW

File: ballista/core/src/physical_optimizer/task_group.rs [hunk 1]
```diff
@@ -110,42 +100,22 @@ impl OptimizeTaskGroup {
...
-    // (so parent nodes won't rely on the output partitioning)
-    if let Some(hash_join) = node.as_any().downcast_ref::<HashJoinExec>() {
-        return matches!(
-            hash_join.output_partitioning(),
-            datafusion::physical_plan::Partitioning::UnknownPartitioning(_)
-        );
-    }
-    false
-}
+// fn is_hash_join_no_partitioning(node: &dyn ExecutionPlan) -> bool {
```
Comment (thinkharderdev) [inline ballista/core/src/physical_optimizer/task_group.rs:108]: @Dandandan Slightly hazy on this but pretty sure we always we want to coalesce after `HashJoinExec` even if it's partitioned.

Comment (Dandandan) [reply ballista/core/src/physical_optimizer/task_group.rs:108]: I think you're right.
The idea was I think we can't pushdown `HashJoinExec` for non-partitioned data, but it should already run in another stage at this ppoint if the child requires a hash-partitioning.

File: ballista/core/src/physical_optimizer/task_group.rs [hunk 2]
```diff
@@ -166,18 +136,29 @@ impl PhysicalOptimizerRule for OptimizeTaskGroup {
     }
 }
 
+/// Return whether we should insert a `CoalesceTasksExec` above this node.
+fn insert_coalesce(plan: &dyn ExecutionPlan) -> bool {
```
Comment (Dandandan) [inline ballista/core/src/physical_optimizer/task_group.rs:140]: I wonder if we're actually 100% correct after this PR. 
If the upstream stage would depend on e.g. distribution (hash partitioning) or sorting being preserved, we could undo the sorting / partitioning by doing the coalescing and writing to the same shuffle files.

Maybe we should an issue which does part of the planning / optimization *before* the required distribution and sorting are satisfied in order to create a correct plan.

**Review by Dandandan — APPROVED** [review]
LGTM, nice cleanup / improvement. Left some comments

**Pull request merged**